### PR TITLE
Golangci lint

### LINKED ISSUE #1640: Consider moving from metalinter to staticcheck
### Description

The next release of @dominikh's [staticcheck](https://github.com/dominikh/go-tools) tool looks very promising:
http://next.staticcheck.io/changes/2018.1

gometalinter, which we're currently, using has significant performance drawbacks, and the complexity of its installation and configuration is also non-trivial. These problems are mostly due to it being a runner of other tools, which staticcheck looks to be solving in its next release. 

If we do end up doing this, let's look into contributing financially to the project - either through sponsorship or purchasing a support package.

### Success Criteria

We should have much faster builds/checks both locally and in CI - possible that we'll also get *better* checks which are consistently updated and added to without us having to manually enable, disable, or configure individual checks.

### PR DESCRIPTION
**Pull request by seebs**
## Overview

This is based on cody's PR #1856, and now includes the vast swathes of minor lint fixes, plus a few actual bug fixes and substantive changes. No doc changes. Should probably update Changelog, etc., but I wanted to get the code out there to be looked at first.

Fixes #1640 

## Pull request checklist

- [X] I have updated the [documentation](https://github.com/pilosa/pilosa/tree/master/docs).
- [X] I have resolved any merge conflicts.
- [X] I have included tests that cover my changes.
- [X] All new and existing tests pass.

## Code review checklist
This is the checklist that the reviewer will follow while reviewing your pull request. You do not need to do anything with this checklist, but be aware of what the reviewer will be looking for.

- [ ] Ensure that any changes to external docs have been included in this pull request.
- [ ] If the changes require that minor/major versions need to be updated, tag the PR appropriately.
- [ ] Ensure the new code is [properly commented](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and follows [Idiomatic Go](https://dmitri.shuralyov.com/idiomatic-go).
- [ ] Check that tests have been written and that they cover the new functionality.
- [ ] Run tests and ensure they pass.
- [ ] Build and run the code, performing any applicable integration testing.

### CODE REVIEW

**Comment by codysoyland** [comment]
I haven't had a chance to review every change, but I like what I've seen so far!

One thing I would suggest is to go ahead and remove the gometalinter job from the CI workflow in this PR but leave it in the Makefile for now.

**Comment by seebs** [comment]
Good idea. (It took me a while to figure out *how*, though, since I don't think I've looked at the CI workflows before, and I wasn't even sure where they were.)

**Comment by seebs** [comment]
Rebased this, fixed new lint which appeared during that time.

File: cmd/root.go [hunk 1]
```diff
@@ -54,9 +54,8 @@ Build Time: ` + pilosa.BuildTime + "\n",
 			if ret, err := cmd.Flags().GetBool("dry-run"); ret && err == nil {
 				if cmd.Parent() != nil {
 					return fmt.Errorf("dry run")
-				} else if err != nil {
-					return fmt.Errorf("problem getting dry-run flag: %v", err)
 				}
+				return fmt.Errorf("problem getting dry-run flag: %v", err)
```
[the remark this answers is not in the document]
Comment (seebs) [reply cmd/root.go]: good catch, it was definitely wrong. gonna check err first, then ret, so it's easier for me to follow.

File: ctl/inspect_test.go [hunk 2]
```diff
@@ -34,14 +34,23 @@ func TestInspectCommand_Run(t *testing.T) {
...
 	}
-	file.Write([]byte("12358267538963"))
+	_, err = file.Write([]byte("12358267538963"))
+	if err != nil {
+		t.Fatalf("writing to tempfile: %v", err)
+	}
 	file.Close()
 	cm.Path = file.Name()
 	err = cm.Run(context.Background())
+	if err != nil && err.Error() != "unmarshalling: reading roaring header: did not find expected serialCookie in header" {
```
[the remark this answers is not in the document]
Comment (seebs) [reply ctl/inspect_test.go]: yes, there is. i should fix the consistency issues, for sure.

on reflection: the test was passing before, but so far as i can tell it's always gotten that error, but possibly i should investigate this one more.

looking at it: we're writing a bunch of raw digits into a file, i don't see why they would be expected to be a valid roaring header. so i think the test is just "are we running inspect on a file at all", not verifying that it actually has valid contents?

File: field_test.go [hunk 3]
```diff
@@ -208,17 +208,20 @@ func TestField_AvailableShards(t *testing.T) {
...
 	// Delete shards; only local shards should remain.
-	f.RemoveAvailableShard(0)
-	f.RemoveAvailableShard(1)
-	f.RemoveAvailableShard(2)
-	f.RemoveAvailableShard(3)
-	f.RemoveAvailableShard(4)
+	for i := uint64(0); i < 5; i++ {
+		err := f.RemoveAvailableShard(i)
+		if err != nil {
+			t.Fatalf("removing shard: %v", err)
```
[the remark this answers is not in the document]
Comment (seebs) [reply field_test.go]: also a good catch

**Comment by seebs** [comment]
rebased on master to try to squash the race failures, which also resulted in one subtle substantive change, and a bunch of changes to the circleci config that were trivial, just immediately next to other lines which changed in master.

**Pull request merged**